File: npmjs/pray-cli/src/cli/materialize.ts
import { existsSync } from "node:fs";
import { resolve } from "node:path";
import { PrayError } from "../errors.js";
import {
  buildLockfile,
  lockfilesEquivalent,
  readLockfile,
  writeLockfileIfChanged,
} from "../lockfile/index.js";
import { defaultLockfilePath, defaultManifestPath } from "../lockfile/paths.js";
import { manifestToJson } from "../manifest/types.js";
import { renderProject, writeRenderedTargets } from "../render/project.js";
import { defaultResolveOptions } from "../resolve/context.js";
import { resolveProject } from "../resolve/project.js";
import { applyPathUpstreamRefreshes } from "../resolve/upstream-refresh.js";
import { runTransaction } from "../transaction/index.js";
import { localSummaryLines } from "./apply-report.js";
import {
  projectRoot,
  resolveCurrentProjectWithGitRefreshFallback,
} from "./invocation.js";

export interface MaterializeOptions {
  manifestPath?: string;
  frozen?: boolean;
  locked?: boolean;
  offline?: boolean;
  refreshSourceRevisions?: boolean;
}

export async function materializeProject(
  options: MaterializeOptions = {},
): Promise<void> {
  return runTransaction(projectRoot(), () => materializeInTransaction(options));
}

async function materializeInTransaction(
  options: MaterializeOptions,
): Promise<void> {
  const manifestPath = resolve(options.manifestPath ?? defaultManifestPath());
  if (!existsSync(manifestPath)) {
    throw PrayError.manifest(
      `missing ${manifestPath}; run pray init to create one`,
    );
  }

  const resolveOptions = {
    ...defaultResolveOptions(),
    offline: options.offline ?? false,
    refreshSourceRevisions: options.refreshSourceRevisions ?? false,
  };
  let project = await resolveCurrentProjectWithGitRefreshFallback(
    resolveOptions,
    !options.locked && !options.frozen,
  );
  const lockfilePath = defaultLockfilePath(project.projectRoot);
  const previousLockfile = existsSync(lockfilePath)
    ? readLockfile(lockfilePath)
    : undefined;
  if (!options.locked && !options.frozen) {
    if (
      await applyPathUpstreamRefreshes(
        project,
        previousLockfile,
        undefined,
        resolveOptions,
      )
    ) {
      project = await resolveCurrentProjectWithGitRefreshFallback(
        resolveOptions,
        true,
      );
    }
  }
  const rendered = renderProject(project);
  const nextLockfile = buildLockfile({
    manifestHash: project.manifestHash,
    ...(project.environment ? { environment: project.environment } : {}),
    projectRoot: project.projectRoot,
    manifestSources: project.manifest.sources,
    manifestTargets: project.manifest.targets,
    rendered,
    packages: project.packages,
    sourceRevisions: project.sourceRevisions,
    sourceHostKeys: project.sourceHostKeys,
    project,
  });

  if (options.frozen || options.locked) {
    const existing = existsSync(lockfilePath)
      ? readLockfile(lockfilePath)
      : undefined;
    if (options.frozen && existing) {
      for (const target of rendered) {
        const outputPath = resolve(project.projectRoot, target.path);
        if (!existsSync(outputPath)) {
          throw PrayError.verify(
            `Rendered file ${target.path} is missing under --frozen`,
          );
        }
      }
    }
    if (options.locked && existing) {
      if (!lockfilesEquivalent(existing, nextLockfile)) {
        throw PrayError.resolution(
          "Prayfile.lock would change under --locked. Run without --locked to update.",
        );
      }
    }
  }

  writeRenderedTargets(project, rendered, previousLockfile);
  writeLockfileIfChanged(lockfilePath, nextLockfile);
  if (!options.frozen && !options.locked) {
    for (const line of localSummaryLines(previousLockfile, project)) {
      process.stdout.write(`${line}\n`);
    }
  }
}

export async function printManifest(
  manifestPath = defaultManifestPath(),
): Promise<void> {
  const project = await resolveProject(manifestPath, defaultResolveOptions());
  process.stdout.write(
    `${JSON.stringify(manifestToJson(project.manifest), null, 2)}\n`,
  );
}

File: rubygems/pray-cli/lib/pray/materialize.rb
# frozen_string_literal: true

module Pray
  module Materialize
    module_function

    def materialize_project(
      manifest_path: nil,
      frozen: false,
      locked: false,
      offline: false,
      refresh: false
    )
      context = Invocation.invocation_context
      manifest_path = File.expand_path(manifest_path || context.manifest_path)
      unless File.exist?(manifest_path)
        raise Error.manifest("missing #{manifest_path}")
      end

      project_root = context.project_root
      options = ResolveOptions.new(
        offline: offline,
        refresh: refresh,
        environment: context.environment
      )
      allow_git_refresh_fallback = !locked && !frozen
      project = begin
        Resolve.resolve_project_in_context(manifest_path, project_root, options)
      rescue Error => error
        if allow_git_refresh_fallback &&
            !offline &&
            !refresh &&
            Resolve.resolution_may_benefit_from_git_source_refresh?(error)
          refreshed = options.dup
          refreshed.refresh = true
          Resolve.resolve_project_in_context(manifest_path, project_root, refreshed)
        else
          raise
        end
      end
      rendered = Render.render_project(project)
      lockfile_path = default_lockfile_path(project.project_root)
      next_lockfile = LockfileIO.build_lockfile(
        project.manifest_hash,
        project.environment,
        project.project_root,
        project.manifest.sources,
        project.manifest.targets,
        rendered,
        project.packages,
        project.source_revisions,
        project.source_host_keys
      )

      if locked
        unless File.exist?(lockfile_path)
          raise Error.verify("missing Prayfile.lock; run install first")
        end
        existing = Pray.read_lockfile(lockfile_path)
        unless Pray.lockfiles_equivalent?(existing, next_lockfile)
          raise Error.verify("lockfile needs update; rerun install to refresh Prayfile.lock")
        end
        unless frozen
          Render.write_rendered_targets(project, rendered)
        end
        if frozen
          rendered.each do |target|
            path = File.join(project.project_root, target.path)
            on_disk = File.read(path)
            if on_disk != target.content
              raise Error.render(
                "#{path} is stale; rerun install to regenerate it or plan to inspect the diff"
              )
            end
          end
        end
        return
      end

      if frozen
        existing = File.exist?(lockfile_path) ? Pray.read_lockfile(lockfile_path) : nil
        if existing
          rendered.each do |target|
            output_path = File.join(project.project_root, target.path)
            unless File.exist?(output_path)
              raise Error.verify("Rendered file #{target.path} is missing under frozen mode")
            end
          end
        end
      end

      Pray.write_lockfile_if_changed(lockfile_path, next_lockfile)
      Render.write_rendered_targets(project, rendered)
    end

    def default_manifest_path(working_directory = Dir.pwd)
      File.join(working_directory, "Prayfile")
    end

    def default_lockfile_path(project_root)
      File.join(project_root, "Prayfile.lock")
    end

    def project_root_from_manifest(manifest_path)
      Resolve.project_root_from_manifest(manifest_path)
    end
  end

  module_function

  def materialize_project(...) = Materialize.materialize_project(...)
  def default_manifest_path(...) = Materialize.default_manifest_path(...)
  def default_lockfile_path(...) = Materialize.default_lockfile_path(...)
  def project_root_from_manifest(...) = Materialize.project_root_from_manifest(...)
  def resolve_project(...) = Resolve.resolve_project(...)
  def render_project(...) = Render.render_project(...)
  def inspect_project(...) = Verify.inspect_project(...)
  def verify_project(...) = Verify.verify_project(...)
  def drift_project(...) = Verify.drift_project(...)
end
